Formatting for time types

markbastian opened the pull request:
Currently, the frontend does not do any special formatting for time of day types. These are always formatted as 12 hour AM/PM times. This PR adds a new function to `frontend/src/metabase/lib/formatting/time.ts`, `formatTimeWithOptions`, that applies user-defined formatting to the time. Previously, `formatTime` was called, which takes no formatting options, so none can be applied.

Fixes #11398

markbastian committed:
Formatting for time types

Currently, the frontend does not do any special formatting for time of day types. These are always formatted as 12 hour AM/PM times. This PR adds a new function to `frontend/src/metabase/lib/formatting/time.ts`, `formatTimeWithOptions`, that applies user-defined formatting to the time. Previously, `formatTime` was called, which takes no formatting options, so none can be applied.

Fixes #11398

markbastian committed:
e2e tests for time of day formatting.

EmmadUsmani commented on frontend/src/metabase/lib/formatting/time.ts:
```diff
@@ -67,3 +67,34 @@ export function formatTimeWithUnit(
+) {
+  const parsedTime = parseTime(time);
+
+  const timeStyle = options.time_style
```
Nit: `options.time_style ?? DEFAULT_TIME_STYLE` is a bit more elegant

EmmadUsmani commented on frontend/src/metabase/lib/formatting/time.ts:
```diff
@@ -67,3 +67,34 @@ export function formatTimeWithUnit(
+    ? options.time_style
+    : DEFAULT_TIME_STYLE;
+
+  const timeEnabled = options.time_enabled
```
Nit: I thought we had a lint rule against nested ternaries since they are hard to read but I guess not. An if-else would be easier to read.

```typescript
  let timeEnabled;
  if (options.time_enabled) {
    timeEnabled = options.time_enabled;
  } else if (hasHour(unit)) {
    timeEnabled = "minute";
  } else {
    timeEnabled = null;
  }
  ```

EmmadUsmani commented on frontend/src/metabase/lib/formatting/time.ts:
```diff
@@ -67,3 +67,34 @@ export function formatTimeWithUnit(
+    ? "minutes"
+    : null;
+
+  const timeFormat = options.time_format
```
Nit: same as before `??` works well here

EmmadUsmani commented on frontend/src/metabase/lib/formatting/value.tsx:
```diff
@@ -177,7 +177,7 @@ export function formatValueRaw(
   } else if (isEmail(column)) {
     return formatEmail(value as string, options);
   } else if (isTime(column)) {
-    return formatTime(value as Moment);
```
We should delete the old `formatTime` entirely I don't think it's used anywhere else, except the static viz `format.ts`, which we probably want to replace with `formatTimeWithOptions` as well

<img width="631" alt="Screenshot 2024-01-19 at 4 11 39 PM" src="https://github.com/metabase/metabase/assets/37751258/25a80112-c2a3-46c0-8f21-c4db4ba3933c">

EmmadUsmani commented on frontend/src/metabase/lib/formatting/time.ts:
```diff
@@ -67,3 +67,34 @@ export function formatTimeWithUnit(
+    ? options.time_format
+    : getTimeFormatFromStyle(timeStyle, unit, timeEnabled as any);
+
+  return parsedTime.format(timeFormat);
```
```suggestion
return parsedTime.isValid() ? parsedTime.format("LT") : String(time);
```

I think we should copy the old return logic from `formatTime`

EmmadUsmani reviewed (commented):
Overall looks good, but I think we can also use the new function in `frontend/src/metabase/static-viz/lib/format.ts` and delete the old `formatTime` function.

markbastian committed:
Update frontend/src/metabase/lib/formatting/time.ts

Co-authored-by: Emmad Usmani <emmadusmani@berkeley.edu>

markbastian committed:
Replacing formatTime with formatTimeWithOptions

This replaces uses of `formatTime` with `formatTimeWithOptions`, updates unit tests, removes references to `formatTime`, and incorporates some code suggestions.

markbastian committed:
Unit tests

markbastian committed:
Adding test id for field info popover.

alxnddr commented on frontend/src/metabase/lib/formatting/time.ts:
```diff
@@ -67,3 +61,35 @@ export function formatTimeWithUnit(
   return m.format(timeFormat);
 }
+
+interface TimeOnlyOptions {
```
Wdyt about moving this type to `frontend/src/metabase/lib/formatting/types.ts`?
I think it would be reasonable to call it `TimeOptions` and extend `OptionsType` with it to avoid duplication

markbastian replied:
So, OptionsType extends from TimeOptions? It seems like other small interfaces have more "locality of behavior" (they live near their functions if only used once), but I don't have strong opinion on this.

markbastian replied:
Done.

alxnddr commented on frontend/src/metabase/lib/formatting/time.ts:
```diff
@@ -67,3 +61,35 @@ export function formatTimeWithUnit(
+  time_style?: string;
+}
+
+export function formatTimeWithOptions(
```
nit: `WithOptions` suffix may be omited since we won't have a separate `formatTime` or `formatTimeWithoutOptions`. Other formatters like `formatUrl` or `formatEmail` accept options and do not have the suffix

markbastian replied:
Good call. Especially since this effectively replaced the old one.

alxnddr commented on e2e/test/scenarios/visualizations-tabular/table.cy.spec.js:
```diff
@@ -373,6 +373,58 @@ describe("scenarios > visualizations > table > conditional formatting", () => {
+      { visitQuestion: true },
+    );
+
+    // Open the formatting menu
```
👍 for clarifying comments

alxnddr reviewed (approved):
Looks great! I left a couple of suggestions, please have a look before merging

markbastian committed:
Refactoring `formatTimeWithOptions` to `formatTime`

markbastian committed:
Moved `TimeOnlyOptions` interface to formatting types and made it an ancestor of `OptionsType`
